File: sol/3342/3342.py
import heapq as hq

class Solution:
    def minTimeToReach(self, moveTime: List[List[int]]) -> int:
        n = len(moveTime)
        m = len(moveTime[0])
        h = [(0, 0, 0, False)]
        visited = [[False] * m for _ in range(n)]
        while len(h) != 0:
            t, x, y, take2Sec = hq.heappop(h)
            if (x, y) == (n - 1, m - 1):
                return t
            if visited[x][y]:
                continue
            visited[x][y] = True
            for dx, dy in ((0, 1), (0, -1), (1, 0), (-1, 0)):
                xx = x + dx
                yy = y + dy
                if 0 <= xx < n and 0 <= yy < m and not visited[xx][yy]:
                    hq.heappush(h, (max(t, moveTime[xx][yy]) + (2 if take2Sec else 1), xx, yy, not take2Sec))
        return -1

File: sol/3342/3342.cpp
#include <queue>

struct Solution {
    static constexpr array<pair<int, int>, 4> DIRS = {{
        {0, 1}, {0, -1}, {1, 0}, {-1, 0}
    }};
    int minTimeToReach(vector<vector<int>>& moveTime) {
        int n = moveTime.size(), m = moveTime[0].size();
        priority_queue<tuple<int, int, int, bool>> h;
        h.push({0, 0, 0, false});
        vector<vector<bool>> visited(n, vector<bool>(m, false));
        while (!h.empty()){
            auto [t, x, y, take2] = h.top();
            h.pop();
            if (x == n - 1 && y == m - 1){
                return -t;
            }
            if (visited[x][y]){
                continue;
            }
            visited[x][y] = true;
            for (auto [dx, dy] : this->DIRS){
                int xx = x + dx, yy = y + dy;
                if (xx >= 0 && xx < n && yy >= 0 && yy < m && !visited[xx][yy]){
                    h.push({
                        min(t, -moveTime[xx][yy]) - (take2 ? 2 : 1),
                        xx, yy, !take2
                    });
                }
            }
        }
        return -1;
    }
};
